Locator/ThemeLocator.php

```diff
@@ -0,0 +1,218 @@
+                '%dir%/%bundle_name%/%override_path%',
+            ),
+        );
+
+        $this->pathPatterns = array_merge_recursive(array_filter($pathPatterns), $defaultPathPatterns);
+    }
+
+    public function locate($theme, $dir = null, $first = true)
+    {
+        // @TODO
+    }
+
+    public function locateThemeInBundles($theme, $dir = null, $first = true)
+    {
+        $parameters = array(
+            '%app_path%' => $this->appPath,
+            '%dir%' => $dir,
+            '%override_path%' => $theme, // ?
+            '%current_theme%' => $theme,
+            '%current_device%' => '', // ?
+            '%template%' => '',
+        );
+
+        foreach ($this->kernel->getBundles() as $bundle) {
+            $checkPaths = $this->getPathsForBundle(
+                array_merge(
+                    $parameters,
+                    array(
+                        '%bundle_path%' => $bundle->getPath(),
+                        '%bundle_name%' => $bundle->getName(),
+                    )
+                )
+            );
+
+            $found = array();
+            foreach ($checkPaths as $checkPath) {
+                if (file_exists($checkPath)) {
+                    if ($first) {
+                        return array(
+                            'path' => $checkPath,
+                            'bundle' => $bundle,
+                        );
+                    }
+                    $found[] = array(
+                        'path' => $checkPath,
+                        'bundle' => $bundle,
+                    );
+                }
+            }
+        }
+
+        if (count($found) > 0 && $first) {
+            return $found[0];
+        }
+
+        return $found;
+    }
+
+    public function locateThemeInApp($theme, $dir = null, $first = true)
+    {
+        $files = array();
+        $parameters = array(
+            '%app_path%' => $this->appPath,
+            '%current_theme%' => $theme,
+            '%current_device%' => '', // ?
+            '%template%' => '',
+        );
+
+        foreach ($this->getPathsForAppResource($parameters) as $checkPath) {
+            if (file_exists($checkPath)) {
+                if ($first) {
+                    return $checkPath;
+                }
+                $files[] = $checkPath;
+            }
+        }
+
+        return $files;
+    }
+
+    /**
+     * Search all themes in paths described in liip_theme config.
+     *     "Goes on adventures, searching for themes".
+     *
+     * @return Array<String> list of found themes.
+     */
+    public function discoverThemes()
+    {
+        $parameters = array(
+            '%app_path%' => $this->appPath,
+            '%dir%' => '',
+            '%override_path%' => '', // ?
+            '%current_theme%' => '',
+            '%current_device%' => '', // ?
+            '%template%' => '',
+        );
+
+        $finder = new Finder();
+        $paths = [];
```

Comment:
as this Bundle is still compatible with PHP 5.3 please do not use short array syntax.